### PR TITLE
(Un)listing of ThingTypes

### LINKED ISSUE #1073: Thing types should have a "listed" flag
With bindings having more and more thing types (newest addition Z-Wave with 100+), it does not make sense to list them all on the "manually add thing" wizard.
Instead, we should have a few more generic thing types being offered there and keep the specific ones only for the auto-discovery and inbox.
I think it should be feasible to simply add a new boolean flag "listed" to the thing type definition and have the UIs filter out the ones that are not meant to be "publicly presented".

### PR DESCRIPTION
**Pull request by sjka**
fixes #1073
Signed-off-by: Simon Kaufmann <simon.kfm@googlemail.com>

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ThingTypeConverter.java [hunk 1]
```diff
@@ -36,6 +37,8 @@
 
     public ThingTypeConverter() {
         super(ThingTypeXmlResult.class, "thing-type");
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ThingTypeConverter.java]: Wouldn't it be simpler to call:
<code>this(ThingTypeXmlResult.class, "thing-type");</code>
So, we do not need to add the two new lines on both constructors.

File: bundles/core/org.eclipse.smarthome.core.thing.xml/thing-description-1.0.0.xsd [hunk 2]
```diff
@@ -35,6 +35,7 @@
             </xs:choice>
         </xs:sequence>
         <xs:attribute name="id" type="config-description:idRestrictionPattern" use="required"/>
+        <xs:attribute name="listed" type="xs:boolean" default="false" use="optional"/>
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing.xml/thing-description-1.0.0.xsd]: Shouldn't the default be set to true?

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/type/BridgeType.java [hunk 3]
```diff
@@ -32,7 +32,7 @@
      */
     public BridgeType(String bindingId, String thingTypeId, String label) throws IllegalArgumentException {
 
-        this(new ThingTypeUID(bindingId, thingTypeId), null, label, null, null, null, null, null);
+        this(new ThingTypeUID(bindingId, thingTypeId), null, label, null, false, null, null, null, null);
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/type/BridgeType.java]: default is true

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/type/ThingType.java [hunk 4]
```diff
@@ -37,12 +38,13 @@
...
     private final Map<String, String> properties;
     private URI configDescriptionURI;
+    private boolean listed;
 
     /**
      * @see ThingType#ThingType(ThingTypeUID, List, String, String, List, List, Map, URI)
      */
     public ThingType(String bindingId, String thingTypeId, String label) throws IllegalArgumentException {
-        this(new ThingTypeUID(bindingId, thingTypeId), null, label, null, null, null, null, null);
+        this(new ThingTypeUID(bindingId, thingTypeId), null, label, null, false, null, null, null, null);
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/type/ThingType.java]: true

File: bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ThingTypeConverter.java [hunk 5]
```diff
@@ -88,12 +93,21 @@ protected ThingTypeConverter(Class clazz, String type) {
...
 
         return thingTypeXmlResult;
     }
 
+    protected boolean getListed(Map<String, String> attributes) {
+        String listedFlag = attributes.get("listed");
+        if (listedFlag != null) {
+            return Boolean.parseBoolean(listedFlag);
+        }
+        return true;
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ThingTypeConverter.java]: Here you implement default = true, in others contexts you are using default = false

File: bundles/model/org.eclipse.smarthome.model.thing.tests/src/org/eclipse/smarthome/model/thing/test/hue/DumbThingTypeProvider.java [hunk 6]
```diff
@@ -31,8 +31,8 @@ public DumbThingTypeProvider() {
         logger.debug("DumbThingTypeProvider created");
         try {
             thingTypes.put(DumbThingHandlerFactory.THING_TYPE_TEST,
-                    new ThingType(DumbThingHandlerFactory.THING_TYPE_TEST, null, "DUMB", "Funky Thing", null, null,
-                            null, new URI("dumb", "DUMB", null)));
+                    new ThingType(DumbThingHandlerFactory.THING_TYPE_TEST, null, "DUMB", "Funky Thing", false, null,
```
Comment (maggu2810) [inline bundles/model/org.eclipse.smarthome.model.thing.tests/src/org/eclipse/smarthome/model/thing/test/hue/DumbThingTypeProvider.java]: true

**Comment by kaikreuzer** [comment]
Could you please also update https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing.xml/thing-description-1.0.0.xsd and https://github.com/eclipse/smarthome/blob/master/docs/documentation/development/bindings/xml-reference.md?

File: bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ThingTypeConverter.java (hunk 5 above)
Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ThingTypeConverter.java]: I fix the others to make it consistent

**Comment by sjka** [comment]
PR has been updated
@kaikreuzer thing-description-1.0.0.xsd was already updated - did I miss anything there?

**Comment by kaikreuzer** [comment]
> thing-description-1.0.0.xsd was already updated - did I miss anything there?

Ok, sorry, I missed that.

PR lgtm.

**Pull request merged**